enhancements/installer/assisted-installer-in-cluster.md

```diff
@@ -0,0 +1,641 @@
+1. The user approves the host for installation. (TODO: how?)
+1. The user or an orchestrator scales up a MachineSet, causing a new Machine resource to be created.
+1. CAPBM binds the Machine to the BareMetalHost, as it does today. It additionally finds the CR representing the agent and uses it to begin installation.
+1. The assisted service initiates installation of the host.
+1. CAPBM updates the status on the BareMetalHost to reflect that it has been provisioned.
+
+#### Day 2 Add Node Virtualmedia Stand-alone
+
+This scenario takes place within a stand-alone bare metal OpenShift cluster.
+
+1. The user creates a BareMetalHost that includes BMC credentials and a label indicating it should be used with assisted installer.
+1. CAPBM gets a URL to the live ISO and adds it to the BareMetalHost, causing it to boot the host.
+1. baremetal-operator uses redfish virtualmedia to boot the live ISO.
+1. The assisted agent starts running on the new hardware and runs through its usual validation and inspection workflow. The wrapping controller creates a new agent resource to be the k8s-native API for the agent.
+1. The user or an orchestrator scales up a MachineSet, resulting in a new Machine being created.
+1. CAPBM does its usual workflow of matching the Machine to an available BareMetalHost. Additionally it uses the agent CR to initiate provisioning of the host.
+1. The assisted service provisions the host.
+
+#### Day 2 Add Node Virtualmedia Multicluster
+
+This scenario takes place from a hub cluster, adding a worker node to a spoke cluster.
+
+1. User creates or modifies a BareMetalAsset on the hub cluster so that its cluster field references the desired ClusterDeployment.
+1. Hive creates a BareMetalHost resource corresponding to the BareMetalAsset and uses it to boot the live ISO.
+1. baremetal-operator uses redfish virtualmedia to boot the live ISO.
+1. Agent running on the host contacts assisted service, resulting in an AssistedInstallAgent resource being created.
+1. Assisted Installer walks the agent through discovery and validation workflows.
+1. When the agent is in a ready state, hive or ACM initiates provisioning, referencing the target cluster's ignition.
+
+#### Create Cluster
+
+This scenario takes place on a hub cluster where hive and possible ACM are
+present. This scenario does not include centralized machine management.
+
+1. User creates BareMetalAssets containing BMC credentials.
+1. User creates AssistedInstallCluster resource describing their desired cluster.
+1. User creates ClusterDeployment describing a new cluster. The Platform section indicates that metal3 should be used to boot hosts with a live ISO. A new field references the AssistedInstallCluster resource.
+1. User associates some BareMetalAssets with the ClusterDeployment.
+1. Hive creates BareMetalHost resources corresponding to each BareMetalAsset, and uses them to boot the live ISO.
```

Comment:
Not sure why this would be Hive, we'll need to discuss.

Reply:
BMA is an ACM API. Hive is not presently aware of it. A dependency from ACM to Hive and from Hive back to ACM feels like trouble, so I'm having a hard time visualizing who or what should be responsible for this automated booting of BMA inventory.

Should BMA CRD move to Hive?

Would the Assisted Install operator be a good place to scan BMA inventory to auto-boot?

Reply:
> Should BMA CRD move to Hive?

That is a key question. I agree with what I think you're implying, which is that this move would make sense if we expect hive to act on BareMetalAssets.

Reply:
> Would the Assisted Install operator be a good place to scan BMA inventory to auto-boot?

One of the advantages of the design in this proposal is that it keeps the assisted installer components out of the business of booting hosts or needing any platform-specific capabilities. It's reasonable to think that other platforms besides bare metal may benefit from this pattern, particularly on-prem virtualization platforms for example. I don't think we would want to oblige assisted installer itself to know how to boot hosts for each platform.